File: Civitas-Java/Civitas/src/main/java/civitas/Tablero.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

/**
 * @author elena
 * @date 20/09/2019
 */

package civitas;

import java.util.ArrayList;

public class Tablero {
    private int numCasillaCarcel = 1;
    private ArrayList<Casilla> casillas;
    private int porSalida;
    private boolean tieneJuez;
    
    //Constructor
    Tablero(int numCasillaCarcel){
        
         if(numCasillaCarcel >= 1)
            this.numCasillaCarcel = numCasillaCarcel;
        
        casillas = new ArrayList();
        Casilla Salida = new Casilla("Salida");
        casillas.add(Salida);
        
        porSalida = 0;
        tieneJuez = false;
    }
    
    private boolean correcto(){
        for(int i = 0; i < casillas.size(); i++){
            if(casillas.get(i).getNombre() == "Juez")
                tieneJuez = true;
        }
        //tieneJuez=true;
        
        if(casillas.size() > this.numCasillaCarcel && tieneJuez)
            return true;        
        
        return false;        
    }
    
    private boolean correcto(int numCasilla){
        if(correcto() && numCasilla < casillas.size())
            return true;
        
        return false;
    }
    
    int getCarcel(){ return this.numCasillaCarcel; }
    
    int getPorSalida(){
        int decremento_salida = porSalida;       
        
        if(porSalida > 0){
            porSalida--;
        }
        
        return decremento_salida;
    }
    
    void añadeCasilla (Casilla casilla){
        
        if(casillas.size() == numCasillaCarcel){
            Casilla carcel = new Casilla("Carcel");
            casillas.add(carcel); 
        }                       
        
        casillas.add(casilla);
        
        if(casillas.size() == numCasillaCarcel){
            Casilla carcel = new Casilla("Carcel");
            casillas.add(carcel);
        }
        
    }
    
    void añadeJuez(){        
        for(int i = 0; i < casillas.size(); i++){
            if(casillas.get(i).getNombre() == "Juez")
                tieneJuez = true;
        }
        
        if(!tieneJuez){
            Casilla juez = new Casilla("Juez");
            casillas.add(juez);
            tieneJuez = true;
        }            
    }
    
    Casilla getCasilla (int numCasilla){        
        if(correcto(numCasilla)){
            return casillas.get(numCasilla);
        }        
        return null;
    }    
    
    int nuevaPosicion (int actual, int tirada){
        int nueva;
        if(!correcto()) return -1;
        
        nueva = (actual + tirada)%casillas.size();
        
        if((actual + tirada) != nueva){
            porSalida++;
        }
        
        return nueva;
    }
    
    int calcularTirada (int origen, int destino){ 
        int tirada = destino - origen;
        
        if(tirada < 0){
            tirada += casillas.size();
        }
        
        return tirada;
    }
}

File: Civitas-Ruby/lib/tablero.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
# encoding:utf-8

require_relative "casilla"

module Civitas
  class Tablero
    def initialize(numCasillaCarcel)
      
      if(numCasillaCarcel >= 1)
        @numCasillaCarcel = numCasillaCarcel
      end
      
      @casillas = Array.new
      salida = Casilla.new_casilla('Salida')
      @casillas << salida 
      
      @porSalida = 0
      @tieneJuez = false
    end
    
    private
    
    def correcto_sin
      @casillas.length > @numCasillaCarcel && @tieneJuez
    end
    
    def correcto(numCasilla)
      correcto_sin && numCasilla < @casillas.length
    end
    
    public
    
    attr_reader :numCasillaCarcel, :porSalida
    
    def getporsalida
      actual = @porSalida
      
      if @porSalida > 0
        @porSalida = @posSalida - 1
      end
      
      return actual
    end
    
    def aniadecasilla(casilla)
            
      if @casillas.length == @numCasillaCarcel
        carcel = Casilla.new_casilla("carcel")
        @casillas << carcel
      end
      
      @casillas << casilla
      
      if @casillas.length == @numCasillaCarcel
        carcel = Casilla.new_casilla("carcel")
        @casillas << carcel
      end
    end
    
    def aniadejuez
      @casillas.each do |casilla|
        if casilla.nombre == 'juez'
          @tieneJuez = true
        end
      end
      
      if !@tieneJuez
        juez = Casilla.new_casillajuez(5, 'juez')
        @casillas << juez
        @tieneJuez = true
      end
    end
    
    def getcasilla(numCasilla)
      if numCasilla >= 0 && numCasilla < @casillas.length
        return @casillas[numCasilla]
      else 
        return nil
      end
    end
    
    def nuevaposicion(actual, tirada)
      if !correcto_sin
        return -1
      end
      
      nueva_posicion = (actual + tirada)%@casillas.length
      
      if nueva_posicion != (actual + tirada)
        @porSalida += 1
      end
      
      return nueva_posicion
    end
    
    def calculartirada(origen, destino)
      tirada = destino - origen
      
      if tirada < 0
        tirada += @casillas.length
      end
      
      return tirada
    end
  end
end
